### PR TITLE
149 release version with plonk support

### LINKED ISSUE #141: Refactor: rename from_reader_unchecked
Rename the `from_reader_unchecked` methods in CircomArkworksPairingBridge etc. to correctly mirror their behavior regarding montgomery reduction.

instead of `from_reader_unchecked` should be `montgomery_from_reader` or similar

### LINKED ISSUE #142: Refactor: common code for co-plonk and co-circom
Some parts of the code are shared between co-plonk and co-groth16. Most notably the SharedWitness, SharedInput and the roots of unity.
As we will do a major refactoring in the near future this is low priority but co-plonk now has a dependency on co-groth16 and we do not want that.

### PR DESCRIPTION
**Pull request by 0xThemis**
doesnt close 140 anymore - we moved it to other PR
closes #141 
closes #142

### CODE REVIEW

**Comment by 0xThemis** [comment]
The updates to the md book will be on a dedicated PR - it is easier to test if the changes were already merged on main

File: circom-snarks/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,192 @@
+#![deny(missing_docs)]
```
Comment (dkales) [inline circom-snarks/src/lib.rs:1]: do we do `deny` or `warn`

File: collaborative-groth16/src/groth16.rs [hunk 2]
```diff
@@ -85,115 +57,14 @@ use g=q^t (this is a 2^s-th root of unity) as (some kind of) generator and compu
 then if log2(domain_size) equals s we take as root of unity q^2, and else we take the log2(domain_size) + 1-th element of the domain created above
 */
 fn root_of_unity_for_groth16<F: PrimeField + FftField>(domain: &GeneralEvaluationDomain<F>) -> F {
-    let (q, roots) = roots_of_unity::<F>();
+    let (q, roots) = circom_snarks::roots_of_unity::<F>();
```
Comment (dkales) [inline collaborative-groth16/src/groth16.rs:59]: this should probably live in a utils module in `circom_snarks`

File: .release-please-manifest.json [hunk 3]
```diff
@@ -3,9 +3,9 @@
   "circom-mpc-vm": "0.2.0",
   "circom-snarks": "0.0.0",
   "circom-types": "0.3.0",
-  "collaborative-circom": "0.2.1",
-  "collaborative-groth16": "0.2.1",
-  "collaborative-plonk": "0.0.0",
+  "co-circom": "0.2.1",
+  "co-groth16": "0.2.1",
+  "co-plonk": "0.0.0",
```
Comment (dkales) [inline .release-please-manifest.json:8]: i don't think we should put 222d147f7a9b5afc71b314d6eb9b4fa423a4c9ff in this PR, and rather sync on it later on in a separate PR tomorrow

**Review by dkales — CHANGES_REQUESTED** [review]

File: Cargo.toml [hunk 4]
```diff
@@ -2,6 +2,7 @@
 members = [
     "circom-mpc-compiler",
     "circom-mpc-vm",
+    "circom-snarks",
```
[the remark this answers is not in the document]
Comment (0xThemis) [reply Cargo.toml:5]: now called co-circom-snarks

File: circom-snarks/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,25 @@
+
+[package]
+name = "circom-snarks"
```
[the remark this answers is not in the document]
Comment (0xThemis) [reply circom-snarks/Cargo.toml:3]: Yeah maybe add one - in this PR?

File: circom-snarks/src/lib.rs (hunk 1 above)
Comment (0xThemis) [reply circom-snarks/src/lib.rs:1]: I think it should be deny ¯\_(ツ)_/¯

But I can change if you think warn is better

File: .release-please-manifest.json (hunk 3 above)
Comment (0xThemis) [reply .release-please-manifest.json:8]: Removed it from the PR

File: co-circom/src/bin/co-circom.rs [hunk 6]
```diff
@@ -624,7 +624,7 @@ where
                             let mut prover = CoGroth16::new(protocol);
 
                             // execute prover in MPC
-                            let proof = prover.prove(&zkey, witness_share)?;
+                            let proof = prover.prove(zkey, witness_share)?;
```
[the remark this answers is not in the document]
Comment (0xThemis) [reply co-circom/src/bin/co-circom.rs:627]: 1) At the moment, Plonk takes the zkey and puts it into a struct called PlonkData, where it lives with the witness therefore we need to take ownership. To mirror the API we did the same in Groth16
2) We do not need the zkey anyways for something other than proving. We do not need to clone it anywhere even if we take ownership.
3) We also consume the witness share, so why not the zkey also?

File: circom-snarks/Cargo.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (dkales) [reply circom-snarks/Cargo.toml:3]: Yeah maybe in a separate commit

File: circom-snarks/src/lib.rs (hunk 1 above)
Comment (dkales) [reply circom-snarks/src/lib.rs:1]: It should just be uniform, I think the others used warn?

File: co-circom/src/bin/co-circom.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (dkales) [reply co-circom/src/bin/co-circom.rs:627]: Just from a library user perspective, having to load/clone the zkey everytime you do the same proof is a lot of overhead. The plonk structure could also store a ref to a zkey internally

File: circom-snarks/src/lib.rs (hunk 1 above)
Comment (0xThemis) [reply circom-snarks/src/lib.rs:1]: I just checked and the other use warn - I changed it to be uniform

File: collaborative-groth16/src/groth16.rs (hunk 2 above)
Comment (0xThemis) [reply collaborative-groth16/src/groth16.rs:59]: now lives under co_circom_snarks::utils

File: co-circom/src/bin/co-circom.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (0xThemis) [reply co-circom/src/bin/co-circom.rs]: OK yeah for the MaxPickChallenge we stores the ZKey but witness was different each time. That makes sense. I will change it - may take a minute or two

**Review by dkales — APPROVED** [review]

**Review by dkales — APPROVED** [review]

**Pull request merged**